### ISSUE TITLE
Add "last modified" properties area to Inspector

### ISSUE THREAD

**Issue by h0lley** [godotengine/godot-proposals#4281]
### Describe the project you are working on

Any project using nodes with long and complex / nested property lists, such as GPUParticles or WorldEnvironment.

### Describe the problem or limitation you are having in your project

It's too cumbersome to navigate long property lists in the inspector. People have expressed that it feels cluttered to them.

With [pinning properties](https://github.com/godotengine/godot-proposals/issues/4274), the user has to exert the additional effort of thinking about which properties to pin and to unpin. In practice, which properties you frequently use changes a lot based on what you are currently working on, so the user would often need to go out of their way to update their pins because they've grown too cluttered.

What's more, it's really hard to predict what properties you are going to need for the next thing you do, so realistically, an user may notice "hey, I am navigating to this property often - I should probably pin it" only after they are halfway through implementing the thing that uses that property.

All in all, I think the effort introduced on the user-side by maintaining pins is just as much if not more than the effort needed to filter or search for a property, so ultimately users will end up not using this feature.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Let's instead have an area at the top of the inspector listing the 5 (amount adjustable in editor settings) properties that have been modified last.

This would be immediately and automatically responsive to the work the user the currently doing without requiring any extra steps on the user-side. It would reduce time spent navigating long property lists by just as much (imho very likely more) than the manual pinning of properties.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Mockup:

![image](https://user-images.githubusercontent.com/23386940/160105140-32e01288-02f8-4eb0-8ee8-9a33acbb8073.png)

Note that Accel Min and Accel Max in this example are properties from a sub-resource. I am not sure if would be helpful if they were highlighted in some way to indicate that. It may not be necessary.

**Ideas for sorting:**

There are multiple ways to sort the last modified properties:
1. Last modified time, the most recently edited proeprty always on top
2. The same order they appear in in the property list to prevent jumping around of properties an preserve some context like Accel Min and Accel Max being listed adjacent to each other
3. Alphabetically - probably an easier to implement alternative to 2.

I'm not certain which one would be best. This could also be an editor setting.

**Potential issues:**

- The "Last modified" area would probably grow flexibly, so it pushes the inspector content beneath it further down when items are added to it. If that's an issue, the scrollbar position should be updated at the same time to account for that. This way, the position of properties would not jump while editing.

- Properties are displayed without the context of their grouping. This can sometimes be confusing; take for example the "Accel Min" property from the mockup. It is actually `linear_accel_min` from the "Linear Accel" group. There is also `radial_accel_min` and `tangential_accel_min` that all show up as "Accel Min" in the inspector. With the current state of the proposal, when the user is unsure, they would need to check the tooltip of the property to get more context. By the way, this issue is much more pronounced with the pinning properties idea, as properties might have been pinned a much longer time ago, so the user is more likely to be needing to look up the missing context.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

n/a

### Is there a reason why this should be core and not an add-on in the asset library?

n/a

**Comment by Zireael07** [comment]
Would be awesome especially when stepping through stuff with the debugger (either in debugger panel or remote view inspector)

<3

**Comment by ScottLongley** [comment]
I'm a big fan of this proposal, however I dislike relying on tooltips so my suggestion for property context would be to allow the user to expand a field revealing the property's path.

![last_modified](https://user-images.githubusercontent.com/29146264/160115871-bfce4803-72cc-45dc-9bba-d5e93c19b27f.png)

**Comment by Mickeon** [comment]
And perhaps double-clicking on the property's name would scroll down to the original path of the property? So any corelated property may become inherently faster to access.

**Comment by arkology** [comment]
Between https://github.com/godotengine/godot-proposals/issues/4274 and this proposal I would prefer this as more user-friendly

**Comment by zinnschlag** [comment]
Strong preference for #4274 here, because it gives the user manual control instead of depending on the most recently used properties; which may be related to what the user wants to do now or it may not.

**Comment by Calinou** [comment]
I also agree with @zinnschlag that this proposal lacks too much user control to really be a net win in all situations.

**Comment by h0lley** [comment]
@zinnschlag I'm sure you can already envision yourself pinning and unpinning those properties like a ninja, but again, I'd predict that you'd stop using this feature after a while because you realize it's more cumbersome than just filtering for properties (for the reasons explained above). In this case, manual control is a bad thing.

How is it not ensured that the properties you modified last aren't also the properties related to what you are currently doing? Sure, you need to navigate to a property the old fashioned way once before it appears in the area for quick access - but that's exactly the same with pinning properties, too. (btw, I think the idea of default pins isn't good either - it's more confusing than anything).

@Calinou are you suggesting that with pinned properties, the property should be hidden from its usual place in the list and exclusively appear as pinned property at the top? I think that's a really bad idea regardless of whether it's been explicitly pinned by the user or not. Imagine a property group where the properties have strong correlation, but since one of them was pinned, it's now missing from the group. That should never happen. It *should* be displayed twice, I don't think that's a problem.

**Comment by Zireael07** [comment]
To add my 3 cents: I am mostly thinking of debugger/remote tree, so I wouldn't change ordering in inspector, just add a highlight to recently changed values.

**Comment by zinnschlag** [comment]
@h0lley Pinning is a feature I would use lightly, because I am actually pretty comfortable with the inspector as it is right now. But I definitely can see use cases for it.

> How is it not ensured that the properties you modified last aren't also the properties related to what you are currently doing? 

People switch tasks. Sometimes rapidly. Sometimes back and forth. Example: You may work on a certain larger task that requires manipulating one or a few properties over many nodes (ideal case for pinning). But there may be regular interruptions that require you also to work on a large number of other properties in between. These are likely to push your main task properties out of the list. I would find that sufficient annoying and disruptive that I wouldn't want to use this feature at all.

And while we are at it: This proposal only covers modifications. What if I only want to look at a property or a few properties of a lot of notes without modifying them? That is not covered at all.

**Comment by h0lley** [comment]
@zinnschlag yea that's a fair point. When working on a larger number of properties, some may be pushed out of the last modified stack too quickly. That's why in hindsight, I added "amount adjustable in editor settings **or perhaps right in the inspector**" to the proposal.

The editor would keep a history of idk, maybe the past 100 changed properties, and we are only peaking at the beginning of that list. If a property is then pushed out, we can temporarily increase the number of visible entries. Perhaps it could look something like this:

![image](https://user-images.githubusercontent.com/23386940/160210537-bfed7896-4c93-4696-9b32-90d00e60c91b.png)

or this:

![image](https://user-images.githubusercontent.com/23386940/160210941-c3340f87-6ad5-413e-899c-91163c35fca1.png)

which is both inspired from the existing UI for editing / browsing arrays. Could even be a combination of both.

In that scenario though, probably only the sorting my last edited time makes any sense.

> And while we are at it: This proposal only covers modifications. What if I only want to look at a property or a few properties of a lot of notes without modifying them? That is not covered at all.

true. I struggle coming up with a common use-case for that though. I think it's a small price to pay if I can have a quick access helper based on what's been modified last rather than having to maintain pins.

**Comment by h0lley** [comment]
@Zireael07 you are thinking of highlighting properties that have been modified last on runtime, is that right? that's an interesting idea, although for this proposal I only had in mind to highlight properties the user manually modified via the inspector as an alternative to the idea of pinning properties.

**Comment by Zireael07** [comment]
@h0lley: Yep, would be awesome for cases like cached resource references or mistakenly flipping a flag on the wrong node...

**Comment by golddotasksquestions** [comment]
I like this proposal, if there is some way in the Editor Settings to optionally hide it. That's because I'm not sure I'd use it all the time and the Inspector column already is often way too long. I'm working on a 1080p laptop and vertical space is therefore sacred.

I don't see a reason why this should not or could not be used in conjunction with the pin proposal.

**Comment by h0lley** [comment]
> I don't see a reason why this should not or could not be used in conjunction with the pin proposal.

I thought that both would be overkill for the very reason you mentioned - lack of space & clutter in the inspector, but if the user could choose which one to enable I guess that wouldn't matter. I still see next to no added value in pinned properties vs. last modified quick access though. They just add complexity in my mind.
